crates/onnx-ir/src/dim_inference.rs

```diff
@@ -128,6 +129,31 @@ fn constant_update_outputs(node: &mut Node) {
     };
 }
 
+fn constant_of_shape_update_output(node: &mut Node) {
+    let value_type = node
+        .attrs
+        .get("value")
+        .map(|v| v.clone().into_tensor().elem_type)
+        .unwrap_or(ElementType::Float32); // If not given, defaults to 0 as float32
+
+    let dim = match &node.inputs[0].ty {
+        // Sometimes the input is a tensor, sometimes it's a shape
+        // because didn't have a chance to update the IR
+        ArgType::Tensor(input_type) => input_type.dim,
+        ArgType::Shape(dim) => *dim,
+        _ => panic!("ConstantOfShape node must have a Tensor or Shape type input"),
+    };
+
+    // Fix the input type to be a shape
+    node.inputs[0].ty = ArgType::Shape(dim);
```

Comment:
This *might* cause an issue depending on the origin  of the input, and how the code generation is handled. 

If `node.inputs[0]` is the output of another node, and that node's output is a tensor, and the `ArgType::Shape(dim)` generates a vector, then passing that will lead to incorrect code generation.

I suspect this is also an issue with the transpose in `onnx-ir/src/coalesce.rs`, and there is two ways I can think of to handle it:
- The easier one in this case is to avoid mutating the input, and check for `ArgType::Tensor` in `burn-import/src/node/costant_of_shape.rs`, and if it is then call `.dim()`.
- The harder, though perhaps better long term, way of handling it is to either rework the parser to inject nodes to do transforms when inputs are mutated, or store/pass a list of arguments where transforms need to happen. I think we can avoid mutating the list of nodes (and thus avoid needing to do an extra pass), since we know the input name that needs to transform by this point in the parsing process

Reply:
@skewballfox yes, I agree the current the input/output type inference is not fully polished up. I will file an issue ticket to refactor for a long term usage. We can apply lessons learned. 

I have a test where an output of Shape node  is Shape arg type. We need to use Shape arg type to declare array types in Rust. So we need to know its size. 

BTW, I was looking at how Netron figures out shapes and it's pretty complicated. Fortunately, for our use case, only ranks and data type matter the most. I think most other DL frameworks track shapes dynamically (may except dfdx). So rank inference will be useful for other frame works.